Lib/test/test_importlib/identity.py

```diff
@@ -0,0 +1,513 @@
+"""
+Provides ``identities_strategy``, a Hypothesis strategy for generating an
+underlying list of identity ``Entries`` with corresponding structured and
+unstructured identity metadata.
+"""
+
+import re
+import typing
+import itertools
+import dataclasses
+
+from test.support.hypothesis_helper import hypothesis
+
+
+@dataclasses.dataclass
+class Token:
+    """
+    Represents a generated identity component. See ``Entry`` for more
+    information.
+    """
+
+    category: typing.Any
+    value: str
+    delim: bool = False
+    ignore: bool = False
+
+
+class Entry(list):
+    """
+    Represent an identity of the form ``(name, email)`` as a list of
+    ``Token`` instances. Delimiter (``delim``) tokens separate adjacent
+    entries or name-email components. Tokens with ``delim`` or ``ignore``
+    have their value omitted from the structured representation.
+    """
+
+    @classmethod
+    def from_empty(cls, length, category, value=""):
+        """
+        Generate an empty identity with ``length`` components - usually
+        two, representing name and email - and ``length - 1`` interleaved
+        delimiters.
+        """
+        tokens = []
+        for index in range(0, length):
+            if index > 0:
+                tokens.append(Token(category, value, True))
+            tokens.append(Token(category, value, False))
+        return cls(tokens)
+
+    def __repr__(self):
+        return type(self).__name__ + "(" + super().__repr__() + ")"
+
+    @property
+    def as_text(self):
+        """
+        Convert to an unstructured textual format - a string. No values
+        are omitted.
+        """
+        return "".join(token.value for token in self)
+
+    @property
+    def as_data(self):
+        """
+        Convert to a structured format - a tuple, split on groups of
+        ``delim`` tokens. Values from ``delim`` and ``ignore`` tokens
+        are omitted. Empty components are converted to ``None``.
+        """
+
+        def keyfunc(token):
+            return token.delim
+
+        data = []
+        for key, group in itertools.groupby(self, key=keyfunc):
+            if key:
+                continue
+            text = "".join(token.value for token in group if not token.ignore)
+            data.append(text if text else None)
+        return tuple(data)
+
+
+@hypothesis.strategies.composite
+def identities_strategy(draw, debug=False):
```

Comment:
These are all ideas I was in the fence about, especially working with strategies vs working with drawn data. I avoided strings because Python wont type check string typos. Perhaps a good middle ground would be predefining some enums.